### PR TITLE
Feature/terminator

### PR DESCRIPTION
**Pull request by denis-yuen**
Compliance work for our back pocket

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java [hunk 1]
```diff
@@ -235,15 +251,25 @@ public boolean selfDestruct(
...
-            cachingAuthenticator.invalidate(token.getContent());
-        }
+    @DELETE
+    @Timed
+    @UnitOfWork
+    @Path("/user/{userId}")
+    @RolesAllowed("admin")
+    @ApiOperation(value = "Terminate user if possible.", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Boolean.class, nickname = "terminateUser")
+    public boolean terminateUser(
+        @ApiParam(hidden = true) @Auth User authUser,  @ApiParam("User to return") @PathParam("userId") long targetUserId) {
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java]: User to ban

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java [hunk 2]
```diff
@@ -235,15 +251,25 @@ public boolean selfDestruct(
...
+        // note this terminates the user but leaves behind a tombstone to prevent re-login
+        checkUser(authUser, authUser.getId());
 
-        return userDAO.delete(user);
+        User targetUser = userDAO.findById(targetUserId);
+        deleteSelfFromEntries(targetUser);
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java]: Should do a null check on targetUser.

**Comment by denis-yuen** [comment]
Reminder to self: pending changes from slack

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java [hunk 3]
```diff
@@ -238,15 +254,28 @@ public boolean selfDestruct(
...
-            // invalidate tokens from caching authenticator
-            cachingAuthenticator.invalidate(token.getContent());
+    @DELETE
+    @Timed
+    @UnitOfWork
+    @Path("/user/{userId}")
+    @RolesAllowed("admin")
+    @ApiOperation(value = "Terminate user if possible.", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Boolean.class, nickname = "terminateUser")
+    public boolean terminateUser(
+        @ApiParam(hidden = true) @Auth User authUser,  @ApiParam("User to ban") @PathParam("userId") long targetUserId) {
```
Comment (garyluu) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java]: I think we should stick to either "Ban" or "Terminate"

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/UserResourceIT.java [hunk 4]
```diff
@@ -97,6 +97,29 @@ public void testChangingNameSuccess() throws ApiException {
...
+        assertFalse(user.getUsername().isEmpty());
+
+        UsersApi adminAdminWebClient = new UsersApi(adminWebClient);
+        final Boolean aBoolean = adminAdminWebClient.terminateUser(user.getId());
+
+        assertTrue(aBoolean);
+
+        boolean shouldFail = false;
+        try {
+            userUserWebClient.getUser();
```
Comment (garyluu) [inline dockstore-integration-testing/src/test/java/io/dockstore/webservice/UserResourceIT.java]: can do
```
try {
            userUserWebClient.getUser();
            assertFail("Should've encountered a something something exception");
        } catch (ApiException e) {
            AssertEqual("blah blah blah", e.getMessage());
        }
```
which avoids the shouldFail

File: dockstore-webservice/src/main/resources/swagger.yaml [hunk 5]
```diff
@@ -3886,6 +3886,29 @@ paths:
...
+      - "users"
+      summary: "Terminate user if possible."
+      description: ""
+      operationId: "terminateUser"
+      produces:
+      - "application/json"
+      parameters:
+      - name: "userId"
+        in: "path"
+        description: "User to return"
```
Comment (garyluu) [inline dockstore-webservice/src/main/resources/swagger.yaml]: Description seems different

**Pull request merged**